{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1dff0536",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import yaml\n",
    "import json\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a3159068",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwl_dir = '/diskmnt/Projects/Users/estorrs/pecgs-somaticwrapper/cwl'\n",
    "Path(cwl_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4ada855a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Step2: Type the coommand line: perl somaticwrapper.pl --srg --sre --wgs --rdir --ref --log --q --mincovt --mincovn --minvaf --maxindsize --exonic --smg --groupname --users --step\n",
    "\n",
    "# rdir = full path of the folder holding files for this sequence run (user must provide)\n",
    "\n",
    "# log = full path of the folder for saving log file; usually upper folder of rdir\n",
    "\n",
    "# srg = bam having read group or not: 1, yes and 0, no (default 1)\n",
    "\n",
    "# sre = re-run: 1, yes and 0, no (default 0)\n",
    "\n",
    "# wgs = 1 if it is wgs data and otherwise it is 0; If you want to output the maf for all variants, set exonic to 0\n",
    "\n",
    "# groupname = job group name: Format; users/groupname\n",
    "\n",
    "# users = user name for job group: Format; users/groupname\n",
    "\n",
    "# step run this pipeline step by step. (user must provide)\n",
    "\n",
    "# ref: the human reference:\n",
    "\n",
    "# q: which queue for submitting job; research-hpc, ding-lab, long (default)\n",
    "\n",
    "# mincovt: minimum coverage for tumor: default >=14\n",
    "\n",
    "# mincovn: minimum coverage for normal: default >=8\n",
    "\n",
    "# minvaf: minimum somatic vaf: default >=0.05\n",
    "\n",
    "# maxindsize: default <= 100\n",
    "\n",
    "# exonic: output exonic region: 1 Yes, 0 No, Default Yes\n",
    "\n",
    "# smg: smg gene list that escapes the 0.05 vaf cut-off\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "0fb108e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my $STRELKA_DIR2=\"/storage1/fs1/songcao/Active/Software/strelka-2.9.10.centos6_x86_64/bin\";\n",
    "# my $pindel=\"/storage1/fs1/songcao/Active/Software/anaconda3/bin/pindel\";\n",
    "# my $PINDEL_DIR=\"/storage1/fs1/songcao/Active/Software/anaconda3/bin\";\n",
    "# my $picardexe=\"/storage1/fs1/songcao/Active/Software/picard/picard.jar\";\n",
    "# my $gatk=\"/storage1/fs1/songcao/Active/Software/GenomeAnalysis/GenomeAnalysisTK.jar\";\n",
    "# my $java_dir=\"/storage1/fs1/songcao/Active/Software/jre1.8.0_121\";\n",
    "# my $java_mutect=\"/storage1/fs1/songcao/Active/Software/jre1.7.0_80\";\n",
    "# my $snpsift=\"/storage1/fs1/songcao/Active/Software/snpEff/20150522/SnpSift.jar\";\n",
    "# my $gatkexe3=\"/storage1/fs1/songcao/Active/Software/gatk/3.7/GenomeAnalysisTK.jar\";\n",
    "# my $mutect1=\"/storage1/fs1/songcao/Active/Software/mutect/mutect-1.1.7.jar\";\n",
    "# my $samtools=\"/storage1/fs1/songcao/Active/Software/samtools/1.2/bin\";\n",
    "# my $samtoolsexe=\"/storage1/fs1/songcao/Active/Software/samtools/1.2/bin/samtools\";\n",
    "# my $varscan=\"/storage1/fs1/songcao/Active/Software/varscan/2.3.8.ndown\";\n",
    "# my $bamreadcount=\"/storage1/fs1/songcao/Active/Software/bam-readcount/0.7.4/bam-readcount\";\n",
    "# my $vepannot=\"/storage1/fs1/songcao/Active/Database/hg38_database/vep/ensembl-tools-release-85/scripts/variant_effect_predictor/variant_effect_predictor.pl\";\n",
    "# my $vepcache=\"/storage1/fs1/songcao/Active/Database/hg38_database/vep/v85\";\n",
    "\n",
    "# my $DB_SNP_NO_CHR=\"/storage1/fs1/songcao/Active/Database/hg38_database/DBSNP/00-All.vcf\";\n",
    "# my $DB_SNP=\"/storage1/fs1/songcao/Active/Database/hg38_database/DBSNP/00-All.chr.vcf\";\n",
    "\n",
    "# my $DB_COSMIC=\"/storage1/fs1/songcao/Active/Database/hg38_database/cosmic/CosmicAllMuts.HG38.sort.chr.vcf\";\n",
    "# my $DB_COSMIC_NO_CHR=\"/storage1/fs1/songcao/Active/Database/hg38_database/cosmic/CosmicAllMuts.HG38.sort.vcf\"; \n",
    "\n",
    "# my $DB_SNP_NO_COSMIC=\"/storage1/fs1/songcao/Active/Database/hg38_database/cosmic/00-All.HG38.pass.cosmic.vcf\";\n",
    "# my $f_ref_annot=\"/storage1/fs1/songcao/Active/Database/hg38_database/vep/Homo_sapiens.GRCh38.dna.primary_assembly.fa\";\n",
    "# my $TSL_DB=\"/storage1/fs1/songcao/Active/Database/tsl/wgEncodeGencodeTranscriptionSupportLevelV23.txt\";\n",
    "# my $h38_REF_bai=$h38_REF.\".fai\";\n",
    "# my $f_gtf= \"/storage1/fs1/songcao/Active/Database/hg38_database/GTF/Homo_sapiens.GRCh38.85.gtf\";\n",
    "\n",
    "# $run_script_path = \"/usr/bin/perl \".$run_script_path.\"/\";\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "db5b9630",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my $STRELKA_DIR2=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/strelka-2.9.10.centos6_x86_64/bin\";\n",
    "# my $pindel=\"pindel\";\n",
    "# my $PINDEL_DIR=\"/storage1/fs1/songcao/Active/Software/anaconda3/bin\";sssssssssssss\n",
    "# my $picardexe=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/picard/picard.jar\";\n",
    "# my $gatk=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/GenomeAnalysis/GenomeAnalysisTK.jar\";\n",
    "# my $java_dir=\"//storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/jre1.8.0_121\";\n",
    "# my $java_mutect=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/jre1.7.0_80\";\n",
    "# my $snpsift=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/snpEff/20150522/SnpSift.jar\";\n",
    "# my $gatkexe3=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/gatk/3.7/GenomeAnalysisTK.jar\";\n",
    "# my $mutect1=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/mutect/mutect-1.1.7.jar\";\n",
    "# my $samtools=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/samtools/1.2/bin\";\n",
    "# my $samtoolsexe=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/samtools/1.2/bin/samtools\";\n",
    "# my $varscan=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/varscan/2.3.8.ndown\";\n",
    "# my $bamreadcount=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/software/bam-readcount/0.7.4/bam-readcount\";\n",
    "# my $vepannot=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/vep/ensembl-tools-release-85/scripts/variant_effect_predictor/variant_effect_predictor.pl\";\n",
    "# my $vepcache=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/vep/v85\";\n",
    "\n",
    "# my $DB_SNP_NO_CHR=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/DBSNP/00-All.vcf\";\n",
    "# my $DB_SNP=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/DBSNP/00-All.chr.vcf\";\n",
    "\n",
    "# my $DB_COSMIC=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/cosmic/CosmicAllMuts.HG38.sort.chr.vcf\";\n",
    "# my $DB_COSMIC_NO_CHR=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/cosmic/CosmicAllMuts.HG38.sort.vcf\"; \n",
    "\n",
    "# my $DB_SNP_NO_COSMIC=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/cosmic/00-All.HG38.pass.cosmic.vcf\";\n",
    "# my $f_ref_annot=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/vep/Homo_sapiens.GRCh38.dna.primary_assembly.fa\";\n",
    "# my $TSL_DB=\"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/tsl/wgEncodeGencodeTranscriptionSupportLevelV23.txt\";\n",
    "# my $h38_REF_bai=$h38_REF.\".fai\";\n",
    "# my $f_gtf= \"/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/somaticwrapper/db/hg38_database/GTF/Homo_sapiens.GRCh38.85.gtf\";\n",
    "\n",
    "# $run_script_path = \"/usr/bin/perl \".$run_script_path.\"/\";\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "605dd373",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = {\n",
    "    'normal_bam': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path.bam',\n",
    "        'secondaryFiles': [\n",
    "            {'class': 'File', 'path': f'a/file/path.bam.bai'},\n",
    "        ]\n",
    "    },\n",
    "    'tumor_bam': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path.bam',\n",
    "        'secondaryFiles': [\n",
    "            {'class': 'File', 'path': f'a/file/path.bam.bai'},\n",
    "        ]\n",
    "    },\n",
    "    'sample': 'a_string',\n",
    "    'reference': {\n",
    "        'class': 'File',\n",
    "        'path': f'a/file/path.fa',\n",
    "        'secondaryFiles': [\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.amb'},\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.ann'},\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.bwt'},\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.fai'},\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.pac'},\n",
    "            {'class': 'File', 'path': f'a/file/path.fa.sa'},\n",
    "            {'class': 'File', 'path': f'a/file/path.dict'}\n",
    "        ]\n",
    "    },\n",
    "    'rescue_genes': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path',\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "4763890a",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(template, open(os.path.join(cwl_dir, 'template.somaticwrapper.yaml'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f8b80c08",
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "cwl = {\n",
    "    'class': 'CommandLineTool',\n",
    "    'cwlVersion': 'v1.0',\n",
    "    '$namespaces': {'sbg': 'https://www.sevenbridges.com/'},\n",
    "    'id': 'somaticwrapper',\n",
    "    'label': 'somaticwrapper',\n",
    "    'baseCommand': ['python', '/pecgs-somaticwrapper/somaticwrapper/somaticwrapper.py'],\n",
    "    'inputs': [\n",
    "        {\n",
    "            'id': 'sample',\n",
    "            'type': 'string',\n",
    "            'inputBinding': {\n",
    "                'position': '1'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'tumor_bam',\n",
    "            'type': 'File',\n",
    "            'secondaryFiles': [\n",
    "               '.bai',\n",
    "            ],\n",
    "            'inputBinding': {\n",
    "                'position': '2'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'normal_bam',\n",
    "            'type': 'File',\n",
    "            'secondaryFiles': [\n",
    "               '.bai',\n",
    "            ],\n",
    "            'inputBinding': {\n",
    "                'position': '3'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'reference',\n",
    "            'type': 'File',\n",
    "            'secondaryFiles': [\n",
    "               '.amb',\n",
    "               '.ann',\n",
    "               '.bwt',\n",
    "               '.fai',\n",
    "               '.pac',\n",
    "               '.sa',\n",
    "               '^.dict'\n",
    "            ],\n",
    "            'inputBinding': {\n",
    "                'prefix': '--reference',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'rescue_genes',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--rescue-genes',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        \n",
    "        # needs path to be set so it works on compute1\n",
    "        {\n",
    "            'id': 'environ_PATH',\n",
    "            'type': 'string?',\n",
    "            'default': '/usr/bin:/opt/vep/src/ensembl-vep:/opt/vep/src/var_c_code:/miniconda/envs/somaticwrapper/bin:$PATH'\n",
    "        },\n",
    "    ],\n",
    "    'arguments': [\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--run-dir',\n",
    "            'valueFrom': 'run'\n",
    "        },\n",
    "    ],\n",
    "    'outputs': [\n",
    "        {\n",
    "            'id': 'dnp_annotated_maf',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'run/run.dnp.annotated.maf'}\n",
    "        },\n",
    "        {\n",
    "            'id': 'withmutect_maf',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'run/run.withmutect.maf'}\n",
    "        },\n",
    "    ],\n",
    "    'requirements': [\n",
    "        {\n",
    "            'class': 'DockerRequirement',\n",
    "            'dockerPull': 'estorrs/pecgs-somaticwrapper:0.0.1'\n",
    "        },\n",
    "        {\n",
    "            'class': 'ResourceRequirement',\n",
    "            'ramMin': 100000\n",
    "        },\n",
    "        {\n",
    "            'class': 'EnvVarRequirement',\n",
    "            'envDef': {\n",
    "                'PATH': '$(inputs.environ_PATH)',\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7e8f979f",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(cwl, open(os.path.join(cwl_dir, 'somaticwrapper.cwl'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c1ad1d3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# t = {\n",
    "#     'class': 'Workflow',\n",
    "#     'cwlVersion': 'v1.0',\n",
    "#     'id': 'somaticwrapper_workflow',\n",
    "#     'label': 'somaticwrapper_workflow',\n",
    "#     'requirements': [{'class': 'ScatterFeatureRequirement'}],\n",
    "#     'inputs': [\n",
    "#         {\n",
    "#             'id': 'sample',\n",
    "#             'type': 'string',\n",
    "#             'inputBinding': {\n",
    "#                 'position': '1'\n",
    "#             }\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'tumor_bam',\n",
    "#             'type': 'File',\n",
    "#             'inputBinding': {\n",
    "#                 'position': '2'\n",
    "#             }\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'normal_bam',\n",
    "#             'type': 'File',\n",
    "#             'inputBinding': {\n",
    "#                 'position': '3'\n",
    "#             }\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'reference',\n",
    "#             'type': 'File',\n",
    "#             'secondaryFiles': [\n",
    "#                '.fai',\n",
    "#             ],\n",
    "#             'inputBinding': {\n",
    "#                 'prefix': '--reference',\n",
    "#                 'position': '0'\n",
    "#             }\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'rescue_genes',\n",
    "#             'type': 'File',\n",
    "#             'inputBinding': {\n",
    "#                 'prefix': '--rescue-genes',\n",
    "#                 'position': '0'\n",
    "#             }\n",
    "#         },\n",
    "#     ],\n",
    "#     'outputs': [\n",
    "#         {\n",
    "#             'id': 'gene_level_cnv',\n",
    "#             'outputSource': 'cnv/gene_level_cnv',\n",
    "#             'type': 'File'\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'arm_level_cnv',\n",
    "#             'outputSource': 'cnv/arm_level_cnv',\n",
    "#             'type': 'File'\n",
    "#         },\n",
    "#     ],\n",
    "#     'steps': [\n",
    "#         {\n",
    "#             'id': 'somaticwrapper',\n",
    "#             'in': [\n",
    "# #                 {'id': 'genome', 'source': 'genome'},\n",
    "# #                 {'id': 'genome_dict', 'source': 'genome_dict'},\n",
    "#                 {'id': 'reference_dir', 'source': 'reference_dir'},\n",
    "#                 {'id': 'target_interval_list', 'source': 'target_interval_list'},\n",
    "#                 {'id': 'common_biallelic', 'source': 'common_biallelic'},\n",
    "#                 {'id': 'protein_coding_gene', 'source': 'protein_coding_gene'},\n",
    "#                 {'id': 'cytoband', 'source': 'cytoband'},\n",
    "#                 {'id': 'pool_of_normals', 'source': 'pool_of_normals'},\n",
    "#                 {'id': 'sample', 'source': 'sample'},\n",
    "#                 {'id': 'normal_bam', 'source': 'stage_normal_bam/output'},\n",
    "#                 {'id': 'tumor_bam', 'source': 'stage_tumor_bam/output'}],\n",
    "#            'label': 'cnv',\n",
    "#            'out': [{'id': 'gene_level_cnv'}, {'id': 'arm_level_cnv'}],\n",
    "#            'run': './cnv.cwl'\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'stage_normal_bam',\n",
    "#             'in': [{'id': 'BAM', 'source': 'normal_bam'}],\n",
    "#             'label': 'stage_normal_bam',\n",
    "#             'out': [{'id': 'output'}],\n",
    "#             'run': './stage_bam.cwl'\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'stage_tumor_bam',\n",
    "#             'in': [{'id': 'BAM', 'source': 'tumor_bam'}],\n",
    "#             'label': 'stage_tumor_bam',\n",
    "#             'out': [{'id': 'output'}],\n",
    "#             'run': './stage_bam.cwl'\n",
    "#         }\n",
    "#     ]\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e91bb557",
   "metadata": {},
   "outputs": [],
   "source": [
    "# yaml.safe_dump(t, open(os.path.join(cwl_dir, 'cnv_workflow.cwl'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d3c1ed6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e49a97d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b43d2fdd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
