{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1dff0536",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import yaml\n",
    "import json\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3159068",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwl_dir = '/diskmnt/Projects/Users/estorrs/pecgs-cnv/cwl'\n",
    "Path(cwl_dir).mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0fb108e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GENOME = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/references/GRCh38.d1.vd1/GRCh38.d1.vd1.fa'\n",
    "# GENOME_DICT = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/references/GRCh38.d1.vd1/GRCh38.d1.vd1.dict'\n",
    "# # REF_DIR = \n",
    "# # TARGET_INTERVAL_LIST = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/db/IDT_xGen_Exome_Research_Panel_v1.hg38.removed_alt_chr.bed.target.preprocessed.exome.interval_list'\n",
    "# TARGET_INTERVAL_LIST = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/db/IDT_xGen_Exome_Research_Panel_v1.hg38.removed_alt_chr.autosome_only.bed.preprocessed.exome.interval_list'\n",
    "# COMMON_BIALLELIC = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/db/snp_database/af-only-gnomad.hg38.common_biallelic.chr1-22XY.vcf'\n",
    "# PROTEIN_CODING_GENE = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/db/gencode.v34.annotation.gene_filtered.protein_coding.ensembl_ID_no_version.protein-coding_hgnc_filtered.duplicates_removed.ensembl_ID_removed.txt'\n",
    "# POOL_OF_NORMALS = '/storage1/fs1/dinglab/Active/Projects/estorrs/pecgs_resources/cnv/pon/create_pon/gatk4scnaPON.Normal.hdf5'\n",
    "# run_cnv(normal_bam, tumor_bam, sample, out_dir, gene_level_script, merge_gene_script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "605dd373",
   "metadata": {},
   "outputs": [],
   "source": [
    "template = {\n",
    "    'normal_bam': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'tumor_bam': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'sample': 'a_string',\n",
    "#     'genome': {\n",
    "#         'class': 'File',\n",
    "#         'path': 'a/file/path'\n",
    "#     },\n",
    "#     'genome_dict': {\n",
    "#         'class': 'File',\n",
    "#         'path': 'a/file/path'\n",
    "#     },\n",
    "    'reference_dir': {\n",
    "        'class': 'Directory',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'target_interval_list': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'common_biallelic': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'protein_coding_gene': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'cytoband': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "    'pool_of_normals': {\n",
    "        'class': 'File',\n",
    "        'path': 'a/file/path'\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4763890a",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(template, open(os.path.join(cwl_dir, 'template.cnv.yaml'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f8b80c08",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwl = {\n",
    "    'class': 'CommandLineTool',\n",
    "    'cwlVersion': 'v1.0',\n",
    "    '$namespaces': {'sbg': 'https://www.sevenbridges.com/'},\n",
    "    'id': 'cnv',\n",
    "    'label': 'cnv',\n",
    "    'baseCommand': ['python', '/pecgs-cnv/cnv/cnv.py'],\n",
    "    'inputs': [\n",
    "        {\n",
    "            'id': 'sample',\n",
    "            'type': 'string',\n",
    "            'inputBinding': {\n",
    "                'position': '1'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'tumor_bam',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'position': '2'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'normal_bam',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'position': '3'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'reference_dir',\n",
    "            'type': 'Directory',\n",
    "\n",
    "        },\n",
    "        {\n",
    "            'id': 'target_interval_list',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--target-interval-list',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'common_biallelic',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--common-biallelic',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'protein_coding_gene',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--protein-coding-gene',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'cytoband',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--cytoband',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        {\n",
    "            'id': 'pool_of_normals',\n",
    "            'type': 'File',\n",
    "            'inputBinding': {\n",
    "                'prefix': '--pool-of-normals',\n",
    "                'position': '0'\n",
    "            }\n",
    "        },\n",
    "        # needs path to be set so it works on compute1\n",
    "        {\n",
    "            'id': 'environ_PATH',\n",
    "            'type': 'string?',\n",
    "            'default': '/miniconda/envs/cnv/bin:$PATH'\n",
    "        },\n",
    "    ],\n",
    "    'arguments': [\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--genome',\n",
    "            'valueFrom': '$(inputs.reference_dir)/GRCh38.d1.vd1.fa'\n",
    "        },\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--genome-dict',\n",
    "            'valueFrom': '$(inputs.reference_dir)/GRCh38.d1.vd1.dict'\n",
    "        },\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--gene-level-script',\n",
    "            'valueFrom': '/pecgs-cnv/cnv/segment_to_geneLevel_v4.py'\n",
    "        },\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--arm-level-script',\n",
    "            'valueFrom': '/pecgs-cnv/cnv/segment_to_chr_arm_level_v4.py'\n",
    "        },\n",
    "        {\n",
    "            'position': 0,\n",
    "            'prefix': '--out-dir',\n",
    "            'valueFrom': 'outputs'\n",
    "        },\n",
    "    ],\n",
    "    'outputs': [\n",
    "        {\n",
    "            'id': 'gene_level_cnv',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'outputs/*T.geneLevel.from_seg.cn'}\n",
    "        },\n",
    "        {\n",
    "            'id': 'arm_level_cnv',\n",
    "            'type': 'File',\n",
    "            'outputBinding': {'glob': 'outputs/*T.bandLevel.from_seg.cn'}\n",
    "        }\n",
    "    ],\n",
    "    'requirements': [\n",
    "        {\n",
    "            'class': 'DockerRequirement',\n",
    "            'dockerPull': 'estorrs/pecgs_cnv:0.0.2'\n",
    "        },\n",
    "        {\n",
    "            'class': 'ResourceRequirement',\n",
    "            'ramMin': 60000\n",
    "        },\n",
    "        {\n",
    "            'class': 'EnvVarRequirement',\n",
    "            'envDef': {\n",
    "                'PATH': '$(inputs.environ_PATH)',\n",
    "            }\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7e8f979f",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(cwl, open(os.path.join(cwl_dir, 'cnv.cwl'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8b126f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# t = yaml.safe_load(open('/home/estorrs/pecgs-pipeline/cwl/msisensor/msisensor_workflow.cwl'))\n",
    "# t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1ad1d3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "t = {\n",
    "    'class': 'Workflow',\n",
    "    'cwlVersion': 'v1.0',\n",
    "    'id': 'cnv_workflow',\n",
    "    'label': 'cnv_workflow',\n",
    "    'requirements': [{'class': 'ScatterFeatureRequirement'}],\n",
    "    'inputs': [\n",
    "        {\n",
    "            'id': 'sample',\n",
    "            'type': 'string',\n",
    "        },\n",
    "        {\n",
    "            'id': 'tumor_bam',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'normal_bam',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'reference_dir',\n",
    "            'type': 'Directory',\n",
    "\n",
    "        },\n",
    "#         {\n",
    "#             'id': 'genome',\n",
    "#             'type': 'File',\n",
    "#         },\n",
    "#         {\n",
    "#             'id': 'genome_dict',\n",
    "#             'type': 'File',\n",
    "#         },\n",
    "        {\n",
    "            'id': 'target_interval_list',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'common_biallelic',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'protein_coding_gene',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'cytoband',\n",
    "            'type': 'File',\n",
    "        },\n",
    "        {\n",
    "            'id': 'pool_of_normals',\n",
    "            'type': 'File',\n",
    "        },\n",
    "    ],\n",
    "    'outputs': [\n",
    "        {\n",
    "            'id': 'gene_level_cnv',\n",
    "            'outputSource': 'cnv/gene_level_cnv',\n",
    "            'type': 'File'\n",
    "        },\n",
    "        {\n",
    "            'id': 'arm_level_cnv',\n",
    "            'outputSource': 'cnv/arm_level_cnv',\n",
    "            'type': 'File'\n",
    "        },\n",
    "    ],\n",
    "    'steps': [\n",
    "        {\n",
    "            'id': 'cnv',\n",
    "            'in': [\n",
    "#                 {'id': 'genome', 'source': 'genome'},\n",
    "#                 {'id': 'genome_dict', 'source': 'genome_dict'},\n",
    "                {'id': 'reference_dir', 'source': 'reference_dir'},\n",
    "                {'id': 'target_interval_list', 'source': 'target_interval_list'},\n",
    "                {'id': 'common_biallelic', 'source': 'common_biallelic'},\n",
    "                {'id': 'protein_coding_gene', 'source': 'protein_coding_gene'},\n",
    "                {'id': 'cytoband', 'source': 'cytoband'},\n",
    "                {'id': 'pool_of_normals', 'source': 'pool_of_normals'},\n",
    "                {'id': 'sample', 'source': 'sample'},\n",
    "                {'id': 'normal_bam', 'source': 'stage_normal_bam/output'},\n",
    "                {'id': 'tumor_bam', 'source': 'stage_tumor_bam/output'}],\n",
    "           'label': 'cnv',\n",
    "           'out': [{'id': 'gene_level_cnv'}, {'id': 'arm_level_cnv'}],\n",
    "           'run': './cnv.cwl'\n",
    "        },\n",
    "        {\n",
    "            'id': 'stage_normal_bam',\n",
    "            'in': [{'id': 'BAM', 'source': 'normal_bam'}],\n",
    "            'label': 'stage_normal_bam',\n",
    "            'out': [{'id': 'output'}],\n",
    "            'run': './stage_bam.cwl'\n",
    "        },\n",
    "        {\n",
    "            'id': 'stage_tumor_bam',\n",
    "            'in': [{'id': 'BAM', 'source': 'tumor_bam'}],\n",
    "            'label': 'stage_tumor_bam',\n",
    "            'out': [{'id': 'output'}],\n",
    "            'run': './stage_bam.cwl'\n",
    "        }\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e91bb557",
   "metadata": {},
   "outputs": [],
   "source": [
    "yaml.safe_dump(t, open(os.path.join(cwl_dir, 'cnv_workflow.cwl'), 'w'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d3c1ed6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e49a97d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b43d2fdd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
